### PR TITLE
Numerical stability of embedding kernels

### PR DESCRIPTION
**Pull request by madsbk**
Address the issue raised in #22377.

The PR #22016 introduces a temporary tensor of weights `grad_weight_per_segment` of the same dtype as the end result, which can be a problem when using `float16`.
In this PR, it now use a `float32` temporary tensor when the input is `float16`.

@ngimel, can I get you to review? I think I have fixed the issues you have pointed out.

### CODE REVIEW

**Review by ngimel — CHANGES_REQUESTED** [review]
there is still partials_per_segment_offset device_vector, and direct access to its elements.

**Comment by madsbk** [comment]
It was my impression that the `policy` in `thrust::exclusive_scan` would make it use the THC allocator? 

The  direct access to the two elements is unfortunately necessary. It need to know the size of `partial_segment_offset` and how many threads to spawn.

**Comment by ngimel** [comment]
device_vector with default allocator template argument https://github.com/pytorch/pytorch/pull/22401/files#diff-5e16509ef3a09789b80e5d16f8dc3062R248 won't use any specified policy, as there's no policy argument in the constructor. 
It's unfortunate about accesses to device memory, if this results in the slowdown for other workloads perhaps the old path will need to be brought back (also it will be needed for cuda graphs later on), but for now it's not a blocker. device_vector is, though.

**Comment by madsbk** [comment]
Make sense, thanks

**Review by ngimel — APPROVED** [review]
approving, hope this will fix @chenyangyu1988 accuracy problems.

File: aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu [hunk 1]
```diff
@@ -13,7 +13,8 @@
 
 #include <thrust/execution_policy.h>
 #include <thrust/unique.h>
-#include <thrust/device_vector.h>
+
+#include <torch/types.h>
```
Comment (mrshenli) [inline aten/src/ATen/native/cuda/EmbeddingBackwardKernel.cu]: @madsbk  I think this file should not include headers under `torch/csrc/api`. If you need to use `torch::kFloat16` and `torch::kFloat32`, you can replace them with `at::kHalf` and `at::kFloat` [[link](https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/types.h#L24-L25)].

**Pull request closed without merging**